BallerinaStreamsV2HoppingWindowTest.testHoppingWindowQuery fails intermittently

**Description:**
```
[ERROR] testHoppingWindowQuery1(org.ballerinalang.stdlib.streams.BallerinaStreamsV2HoppingWindowTest)  Time elapsed: 6.366 s  <<< FAILURE!
java.lang.AssertionError: Expected events are not received expected [2] but found [1]
	at org.ballerinalang.stdlib.streams.BallerinaStreamsV2HoppingWindowTest.testHoppingWindowQuery1(BallerinaStreamsV2HoppingWindowTest.java:51)
```

--- a/stdlib/streams/src/main/ballerina/streams/windows.bal
+++ b/stdlib/streams/src/main/ballerina/streams/windows.bal
@@ -1801,10 +1801,11 @@ public type HoppingWindow object {
     public int timeInMilliSeconds;
     public int hoppingTime;
     public any[] windowParameters;
+    public int nextEmitTime = -1;
     public LinkedList currentEventQueue;
+    public LinkedList? expiredEventQueue;
     public StreamEvent? resetEvent;
     public task:Timer? timer;
-    public boolean isStart;
     public function (StreamEvent?[])? nextProcessPointer;
 
     public function __init(function (StreamEvent?[])? nextProcessPointer, any[] windowParameters) {
@@ -1815,7 +1816,7 @@ public type HoppingWindow object {
         self.resetEvent = ();
         self.timer = ();
         self.currentEventQueue = new();
-        self.isStart = true;
+        self.expiredEventQueue = ();
         self.initParameters(self.windowParameters);
     }
 
@@ -1846,55 +1847,60 @@ public type HoppingWindow object {
     public function invokeProcess() returns error? {
         map<anydata> data = {};
         StreamEvent timerEvent = new(("timer", data), "TIMER", time:currentTime().time);
-        StreamEvent?[] timerEventWrapper = [timerEvent];
+        StreamEvent?[] timerEventWrapper = [];
+        timerEventWrapper[0] = timerEvent;
         self.process(timerEventWrapper);
         return ();
     }
 
     public function process(StreamEvent?[] streamEvents) {
         LinkedList outputStreamEvents = new();
-        if (self.isStart) {
+        if (self.nextEmitTime == -1) {
+            self.nextEmitTime = time:currentTime().time + self.hoppingTime;
             self.timer = new
             task:Timer(function () returns error? {return self.invokeProcess();},
-                function (error e) {self.handleError(e);}, self.hoppingTime, delay = self.hoppingTime);
+                function (error e) {self.handleError(e);}, self.hoppingTime);
             _ = self.timer.start();
-            self.isStart = false;
         }
 
-        LinkedList streamEventChunk = new;
-        lock {
-            foreach var event in streamEvents {
-                streamEventChunk.addLast(event);
+        int currentTime = time:currentTime().time;
+        boolean sendEvents = false;
+
+        if (currentTime >= self.nextEmitTime) {
+            self.nextEmitTime += self.hoppingTime;
+            self.timer.stop();
+            self.timer = new
+            task:Timer(function () returns error? {return self.invokeProcess();},
+                function (error e) {self.handleError(e);},
+                self.hoppingTime);
+            _ = self.timer.start();
+            sendEvents = true;
+        } else {
+            sendEvents = false;
+        }
+
+        foreach var evt in streamEvents {
+            StreamEvent event = <StreamEvent> evt;
+            if (event.eventType != "CURRENT") {
+                continue;
             }
-
-            streamEventChunk.resetToFront();
-
-            while (streamEventChunk.hasNext()) {
-                StreamEvent streamEvent = <StreamEvent>streamEventChunk.next();
-
-                if (streamEvent.eventType == "CURRENT") {
-                    StreamEvent clonedEvent = streamEvent.copy();
-                    self.currentEventQueue.addLast(clonedEvent);
-                } else {
-                    self.currentEventQueue.resetToFront();
-                    while (self.currentEventQueue.hasNext()) {
-                        StreamEvent currEvent = getStreamEvent(self.currentEventQueue.next());
-                        if (currEvent.timestamp >= (streamEvent.timestamp - self.timeInMilliSeconds)) {
-                            if (self.resetEvent is ()) {
-                                self.resetEvent = createResetStreamEvent(currEvent);
-                            }
-                            outputStreamEvents.addLast(currEvent);
-                        } else {
-                            self.currentEventQueue.removeCurrent();
-                            continue;
-                        }
-                        if (currEvent.timestamp < (streamEvent.timestamp + self.hoppingTime - self.timeInMilliSeconds)) {
-                            self.currentEventQueue.removeCurrent();
-                        }
-                    }
-                    if (!(self.resetEvent is ())) {
-                        outputStreamEvents.addFirst(self.resetEvent);
-                        self.resetEvent = ();
+            StreamEvent clonedEvent = event.copy();
+            self.currentEventQueue.addLast(clonedEvent);
+        }
+        if (sendEvents) {
+            if (self.currentEventQueue.getFirst() != ()) {
+                if (!(self.resetEvent is ())) {
+                    outputStreamEvents.addLast(self.resetEvent);
+                    self.resetEvent = ();
+                }
+                self.resetEvent = createResetStreamEvent(getStreamEvent(self.currentEventQueue.getFirst()));
+                self.currentEventQueue.resetToFront();
+                while (self.currentEventQueue.hasNext()) {
+                    StreamEvent streamEvent = getStreamEvent(self.currentEventQueue.next());
+                    if (streamEvent.timestamp >= currentTime - self.timeInMilliSeconds) {
+                        outputStreamEvents.addLast(streamEvent);
+                    } else {
+                        self.currentEventQueue.removeCurrent();
                     }
                 }
             }

